test/integration/core/staking-limits.integration.ts

```diff
@@ -0,0 +1,232 @@
+    await expect(lido.connect(stranger).removeStakingLimit()).to.be.revertedWith("APP_AUTH_FAILED");
+
+    await lido.connect(voting).removeStakingLimit();
+  });
+
+  it("Should allow staking above limit after removal", async () => {
+    await lido.connect(voting).setStakingLimit(ether("1"), ether("0.01"));
+    await lido.connect(voting).removeStakingLimit();
+
+    await lido.connect(stranger).submit(ethers.ZeroAddress, { value: ether("10") });
+  });
+
+  it("Should prevent staking when protocol is stopped", async () => {
+    await lido.connect(voting).stop();
+
+    await expect(lido.connect(stranger).submit(ethers.ZeroAddress, { value: ether("1") })).to.be.revertedWith(
+      "STAKING_PAUSED",
+    );
+
+    await lido.connect(voting).resume();
+  });
+
+  it("Should mint correct stETH amount when staking", async () => {
+    const stakeAmount = ether("1");
+
+    const balanceBefore = await lido.balanceOf(stranger.address);
+    await lido.connect(stranger).submit(ethers.ZeroAddress, { value: stakeAmount });
+    const balanceAfter = await lido.balanceOf(stranger.address);
+
+    expect(balanceAfter - balanceBefore).to.be.gte(stakeAmount - 2n);
+  });
+
+  const testCases = [
+    { maxLimit: 10n ** 6n, limitPerBlock: 10n ** 4n },
+    { maxLimit: 10n ** 12n, limitPerBlock: 10n ** 10n },
+    { maxLimit: 10n ** 18n, limitPerBlock: 10n ** 16n },
+  ];
+
+  for (const { maxLimit, limitPerBlock } of testCases) {
+    it(`Should update staking limits correctly with max=${maxLimit}, perBlock=${limitPerBlock}`, async () => {
+      const localSnapshot = await Snapshot.take();
+
+      // Set staking limit
+      await lido.connect(voting).setStakingLimit(maxLimit, limitPerBlock);
+
+      // Get initial stake limit
+      const stakingLimitBefore = await lido.getCurrentStakeLimit();
+      expect(stakingLimitBefore).to.equal(maxLimit);
+
+      // Submit stake
+      await lido.connect(stranger).submit(ethers.ZeroAddress, { value: limitPerBlock });
+
+      // Check limit decreased by submitted amount
+      const stakingLimitAfterSubmit = await lido.getCurrentStakeLimit();
+      expect(stakingLimitAfterSubmit).to.equal(stakingLimitBefore - limitPerBlock);
+
+      await mine(1);
+
+      // Check limit restored to max
+      const stakingLimitAfterBlock = await lido.getCurrentStakeLimit();
+      expect(stakingLimitAfterBlock).to.equal(stakingLimitBefore);
+
+      await Snapshot.restore(localSnapshot);
+    });
+  }
+
+  context("Staking limit events", () => {
+    it("Should emit correct event when setting staking limit", async () => {
+      const maxLimit = ether("1000");
+      const limitPerBlock = ether("100");
+
+      const tx = await lido.connect(voting).setStakingLimit(maxLimit, limitPerBlock);
+      const receipt = await tx.wait();
+
+      expect(receipt?.logs.length).to.equal(1);
+      await expect(tx).to.emit(lido, "StakingLimitSet").withArgs(maxLimit, limitPerBlock);
```

Comment:
Maybe `await expect(lido.connect(voting).setStakingLimit(maxLimit, limitPerBlock)).to.emit...` we don't use tx here except for events checking, but can be simplified

Reply:
And below also